### PR TITLE
Controller for the Pod Info Custom Resource

### PR DESCRIPTION
**Pull request by prateek041**
This PR contains:

- Controller logic for reconciling the PodInfo custom Resource.
- Unit and Integration tests for the controller.
- Logic to run the controller using tetragon operator binary.
- CI Workflow to test PodInfo controller.

### CODE REVIEW

**Comment by michi-covalent** [comment]
let's make this a draft while we fix the CI failures

**Comment by prateek041** [comment]
There is Name mismatch in helm charts for now. Raised an issue for that here #1421 .

Make that change locally and run `make run-tests`, to test controller. I made some changes in the logic today and manually tested, works correctly. main focus is testing it now.

cc @michi-covalent

File: operator/Makefile
Comment (michi-covalent) [inline operator/Makefile:1]: let's clean up this Makefile and only keep the targets we actually use.

File: operator/bin/setup-envtest
Comment (michi-covalent) [inline operator/bin/setup-envtest:1]: i didn't realize you can do `UseExistingCluster` with envtest. let's just use kind, we can get rid of this file.

File: operator/cmd/serve/serve.go
Comment (michi-covalent) [inline operator/cmd/serve/serve.go:29]:
```suggestion
```

File: operator/cmd/serve/serve.go
Comment (michi-covalent) [inline operator/cmd/serve/serve.go:43]:
```suggestion
			log := logrusr.New(logging.DefaultLogger.WithField(logfields.LogSubsys, "operator"))
```

File: operator/main.go
Comment (michi-covalent) [inline operator/main.go:12]: nit delete line 12

File: operator/podinfo/podinfo_controller.go
Comment (michi-covalent) [inline operator/podinfo/podinfo_controller.go:43]: curious what happens if we don't have this logic? doesn't k8s garbage collect the pod info when the owner gets deleted?

File: operator/podinfo/podinfo_controller_suite_test.go
Comment (michi-covalent) [inline operator/podinfo/podinfo_controller_suite_test.go:42]: please address these TODOs 🙏

File: operator/podinfo/podinfo_controller_test.go
Comment (michi-covalent) [inline operator/podinfo/podinfo_controller_test.go:90]:
```suggestion
```

File: operator/podinfo/podinfo_controller_test.go
Comment (michi-covalent) [inline operator/podinfo/podinfo_controller_test.go:325]:
```suggestion
```

**Review by michi-covalent — COMMENTED** [review]
we also need to add a github action workflow file similar to https://github.com/cilium/tetragon/blob/main/.github/workflows/gotests.yml that runs these tests on each PR / push.

File: operator/podinfo/podinfo_controller.go
Comment (prateek041) [reply operator/podinfo/podinfo_controller.go:43]: that is a good point, Initially I planned to handle as many cases as possible and then incrementally remove/improve things.

Let me try and remove this and see if deletetion still works

File: operator/podinfo/podinfo_controller_suite_test.go
Comment (prateek041) [reply operator/podinfo/podinfo_controller_suite_test.go:42]: yes, I am almost done with the logging, just haven't pushed it because it's not clean enough yet. I should do it first.

File: operator/podinfo/podinfo_controller_test.go
Comment (prateek041) [reply operator/podinfo/podinfo_controller_test.go:325]: Commented it out, because I found a corner case, need to fix that and uncomment.

File: operator/bin/setup-envtest
Comment (prateek041) [reply operator/bin/setup-envtest:1]: Sure.

File: operator/main.go
Comment (prateek041) [reply operator/main.go:12]: done

File: operator/podinfo/podinfo_controller_test.go
Comment (prateek041) [reply operator/podinfo/podinfo_controller_test.go:90]: fixed

File: operator/Makefile
Comment (prateek041) [reply operator/Makefile:1]: done

File: operator/cmd/serve/serve.go
Comment (prateek041) [reply operator/cmd/serve/serve.go:29]: done

File: operator/cmd/serve/serve.go
Comment (prateek041) [reply operator/cmd/serve/serve.go:43]: done

File: operator/podinfo/podinfo_controller_suite_test.go
Comment (prateek041) [reply operator/podinfo/podinfo_controller_suite_test.go:42]: done

**Comment by michi-covalent** [comment]
let me rebase and squash.

File: operator/podinfo/podinfo_controller.go
Comment (michi-covalent) [inline operator/podinfo/podinfo_controller.go:262]: try setting `Controller` field to true

**Comment by michi-covalent** [comment]
@prateek041 just a heads up i'm updating this PR 📝

File: operator/podinfo/podinfo_controller.go
Comment (lambdanis) [inline operator/podinfo/podinfo_controller.go:91]: Can we use reflect.DeepEqual on `OwnerReferences`? Or, if it's problematic, extract this part to a separate function maybe. I got confused for a moment why all these checks are here.

File: operator/podinfo/podinfo_controller.go
Comment (lambdanis) [inline operator/podinfo/podinfo_controller.go:23]: I don't see it being used, is it needed?

**Review by lambdanis — COMMENTED** [review]
I left a few comments, or rather questions. I haven't reviewed the tests yet, will do tomorrow.

Thank you @prateek041 for this PR, I'm looking forward to this feature.

File: operator/cmd/serve/serve.go
Comment (prateek041) [reply operator/cmd/serve/serve.go:36]: Deleting makes more sense.

File: operator/podinfo/podinfo_controller.go
Comment (prateek041) [reply operator/podinfo/podinfo_controller.go:2]: I was not sure what to put there (Not Familiar with Semantics), what should I put there ?

File: operator/podinfo/podinfo_controller.go
Comment (prateek041) [reply operator/podinfo/podinfo_controller.go:91]: I am not sure if we can use DeepEqual here because we are comparing two separate things (Pod and PodInfo). I can separate out the logic for ownerReference.

Comment (lambdanis) [reply operator/podinfo/podinfo_controller.go:91]: I see. Yeah, `reflect.DeepEqual` will always return false if we compare Pod and PodInfo, but we can use it to compare only OwnerReferences, which are same type in both I believe.

File: operator/podinfo/podinfo_controller.go
Comment (lambdanis) [reply operator/podinfo/podinfo_controller.go:2]: I would expect all files to have
```
// Copyright Authors of Tetragon
```

File: operator/podinfo/podinfo_controller.go
Comment (prateek041) [reply operator/podinfo/podinfo_controller.go:91]: sure, let me try both and see what works. I am a bit unsure because `podInfo` are owned by `Pods`, not the other way around, so their OwnerReferences might not match.

Comment (lambdanis) [reply operator/podinfo/podinfo_controller.go:91]: Ah, sorry, I made it confusing. Yes, OwnerReferences will be very different for Pod and PodInfo. We should not compare these two, but compare PodInfo OwnerReferenes to an OwnerReferences literal (as you did in this PR, just I would use a different syntax for that).

Comment (prateek041) [reply operator/podinfo/podinfo_controller.go:91]: Not a problem ! So, will a separate function work ?

File: operator/podinfo/podinfo_controller.go
Comment (michi-covalent) [reply operator/podinfo/podinfo_controller.go:82]: these are maps though right? for podIPs yes let's use slices.Equal 🥰

File: operator/podinfo/podinfo_controller.go
Comment (michi-covalent) [reply operator/podinfo/podinfo_controller.go:91]: i think what anna is suggesting is to define an expected owners reference like this:
```
	controller := true
	blockOwnerDeletion := true
	expectedOwnerReference := metav1.OwnerReference{
		APIVersion:         "v1",
		Kind:               "Pod",
		Name:               pod.Name,
		UID:                pod.UID,
		Controller:         &controller,
		BlockOwnerDeletion: &blockOwnerDeletion,
	}
```
and then compare it against podinfo owner reference using `reflect.DeepEqual`.

File: operator/podinfo/podinfo_controller.go
Comment (lambdanis) [reply operator/podinfo/podinfo_controller.go:82]: Ah. Of course these are maps. [maps.Equal](https://pkg.go.dev/maps#Equal) then?

Comment (michi-covalent) [reply operator/podinfo/podinfo_controller.go:82]: how did i not know about this package 😹 yeah maps.Equal sounds good to me 👍

Comment (prateek041) [reply operator/podinfo/podinfo_controller.go:82]: What is the benefit of using `slices.Equal` here ?

Also, `slices.Equal` expects two `[]string` but the PodIPs are `[]PodIP`. To use slices.Equal, I need to convert it to string type, so, should I do that ?

File: operator/podinfo/podinfo_controller.go
Comment (prateek041) [reply operator/podinfo/podinfo_controller.go:23]: yes, while setting up the controller, the client and scheme is being assigned. In the `cmd/serve/serve.go`

```Go
if err = (&podinfo.Reconciler{
	Client: mgr.GetClient(),
	Scheme: mgr.GetScheme(),
}).SetupWithManager(mgr); err != nil {
	setupLog.Error(err, "unable to create controller", "controller", "podinfo")
	os.Exit(1)
```

Here: https://github.com/cilium/tetragon/pull/1410/files#diff-a4c1a3f9a379da52152a1e2f245c40c5c541b3af8243cf61a66554a15f4d7d52R68-R74

File: operator/podinfo/podinfo_controller.go
Comment (michi-covalent) [reply operator/podinfo/podinfo_controller.go:82]: `slices.Equal` is a generic function, so it works so long as the types match.

... and now i remember why we didn't use `slices.Equal`. `PodIP` type in `PodInfo` is different from `PodIP` type in `Pod`, so `slices.Equal` complains:

```
Cannot use 'podInfo.Status.PodIPs' (type []PodIP) as the type S ([]PodIP)
```

so keep the code as is for `PodIPs` comparison, and use `maps.Equal` for annotations and labels.

Comment (prateek041) [reply operator/podinfo/podinfo_controller.go:82]: sure, annotations and labels ✅

File: operator/cmd/serve/serve.go
Comment (lambdanis) [inline operator/cmd/serve/serve.go:49]: One more question for my education: where does this value come from? Is `f161f714` just a random prefix?

File: operator/podinfo/podinfo_controller.go
Comment (lambdanis) [reply operator/podinfo/podinfo_controller.go:24]: I see, I missed it earlier, thanks. But I still don't see this value read anywhere, am I missing something?

File: operator/podinfo/podinfo_controller_test.go
Comment (lambdanis) [inline operator/podinfo/podinfo_controller_test.go:140]:
```suggestion
		assert.Equal(t, expectedPodInfo, generatedPodInfo, "Generated incorrect PodInfo corresponding to the pod")
```
typo

File: operator/podinfo/podinfo_controller_test.go
Comment (lambdanis) [inline operator/podinfo/podinfo_controller_test.go:59]:
```suggestion
```
this looks like a leftover

File: operator/podinfo/podinfo_controller_test.go
Comment (lambdanis) [inline operator/podinfo/podinfo_controller_test.go:106]:
```suggestion
// TestGeneratePod tests if the corresponding podInfo CR generated for a pod is same when using generatePodInfo function
```

**Review by lambdanis — COMMENTED** [review]
@prateek041 It looks like you need to rebase the branch due to conflicts in `go.mod`.

Looks good to me overall. I found some typos in tests, nice tests coverage :+1:

File: operator/podinfo/podinfo_controller.go
Comment (michi-covalent) [reply operator/podinfo/podinfo_controller.go:24]: yeah i think we just copied this from somewhere, and i don't know how or if it actually gets used. @prateek041 do you wanna try removing the `Scheme` field and see what happens? 🧑‍🔬

Comment (michi-covalent) [reply operator/podinfo/podinfo_controller.go:24]: most likely we don't need it because we don't call these functions:

- https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#SetControllerReference
- https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#SetOwnerReference

Comment (prateek041) [reply operator/podinfo/podinfo_controller.go:24]: sure ! let me take a look at this.

File: operator/cmd/serve/serve.go
Comment (prateek041) [reply operator/cmd/serve/serve.go:49]: I am not sure about this. @michi-covalent can better answer this.

Comment (prateek041) [reply operator/cmd/serve/serve.go:49]: So, should it be changed ? or I should leave it the way it is and resolve this conversation ?

Comment (michi-covalent) [reply operator/cmd/serve/serve.go:49]: yeah just leave it as is

File: operator/podinfo/podinfo_controller.go
Comment (prateek041) [reply operator/podinfo/podinfo_controller.go]: I removed the scheme, the tests still pass. I guess we did not needed it 😄

**Review by michi-covalent — APPROVED** [review]
let's wait for an ack from anna before merging ☺️

**Review by lambdanis — APPROVED** [review]

**Comment by michi-covalent** [comment]
got a verbal approval from anna. shipping 🚀

**Pull request merged**
